analytics/src/main/java/com/segment/analytics/internal/AnalyticsClient.java

```diff
@@ -227,22 +263,22 @@ public void run() {
               log.print(VERBOSE, "Flushing messages.");
             }
           } else {
+            messageIdSizeMap.put(message.messageId(), messageSizeInBytes(message));
             messages.add(message);
           }
 
           Boolean isBlockingSignal = message == FlushMessage.POISON || message == StopMessage.STOP;
           Boolean isOverflow = messages.size() >= size;
 
           if (!messages.isEmpty() && (isOverflow || isBlockingSignal)) {
-            Batch batch = Batch.create(CONTEXT, messages);
+            Batch batch = BatchUtility.createBatch(messages, messageIdSizeMap, contextSize);
```

Comment:
I don't think we need this method at all, if we do the check before adding the message to the queue. try the following logic see if that makes sense
```
precompute the total size with the new message

if overflow
    snapshot messages list
    create new messages list
    create batch (Batch.create(CONTEXT, snapshotted messages)
    create upload task

add message to the queue
```